http3/client.go

```diff
@@ -98,32 +110,92 @@ func (c *client) dial() error {
 	}
 
 	// run the sesssion setup using 0-RTT data
-	go func() {
-		if err := c.setupSession(); err != nil {
-			c.logger.Debugf("Setting up session failed: %s", err)
-			c.session.CloseWithError(quic.ErrorCode(errorInternalError), "")
-		}
-	}()
+	// go func() {
+	if rerr := c.setupSession(); rerr.HasError() {
+		c.logger.Debugf("Setting up session failed: %s", rerr)
+		c.session.CloseWithError(quic.ErrorCode(rerr.connErr), "")
+	}
+	// }()
 
 	return nil
 }
 
-func (c *client) setupSession() error {
+func (c *client) setupSession() requestError {
 	// open the control stream
 	str, err := c.session.OpenUniStream()
 	if err != nil {
-		return err
+		return newConnError(errorMissingSettings, err)
 	}
 	buf := &bytes.Buffer{}
 	// write the type byte
 	buf.Write([]byte{0x0})
 	// send the SETTINGS frame
 	(&settingsFrame{}).Write(buf)
 	if _, err := str.Write(buf.Bytes()); err != nil {
-		return err
+		return newConnError(errorMissingSettings, err)
 	}
 
-	return nil
+	controlStreamIn, err := c.session.AcceptUniStream(context.Background())
+	if err != nil {
+		c.logger.Debugf("Accepting the incoming control stream failed.")
+		return newConnError(errorInternalError, err)
+	}
+
+	br, ok := controlStreamIn.(byteReader)
+	if !ok {
+		br = &byteReaderImpl{controlStreamIn}
+	}
+	t, err := utils.ReadVarInt(br)
+	if t != 0x0 {
+		c.logger.Debugf("First stream must be a control stream")
+		return newConnError(errorMissingSettings, err)
+	}
+
+	frame, err := parseNextFrame(controlStreamIn)
+	if err != nil {
+		c.logger.Debugf("Error encountered while parsing incoming frame")
+		return newConnError(errorStreamCreationError, err)
+	}
+	sf, ok := frame.(*settingsFrame)
+	if !ok {
+		c.logger.Debugf("First incoming frame parsed was not a settings frame")
+		return newConnError(errorMissingSettings, nil)
+	}
+	// TODO: do something with the settings frame
+	c.logger.Debugf("Got settings frame: %+v", sf)
+	go func() {
+		for {
+			frame, err := parseNextFrame(controlStreamIn)
+			if err != nil {
+				// Hack
+				if err.Error() == "Application error 0x100" {
+					c.logger.Debugf("Closing control streams")
+					return
+				}
+				c.logger.Debugf("Error encountered while parsing incoming frame: %s", err)
+				return
+			}
+			switch f := frame.(type) {
+			case *goawayFrame:
+				c.logger.Debugf("Received goaway frame, halting requests with streamID >= %d", f.StreamID)
+				c.clientCancel()
+				c.mutex.Lock()
+				c.logger.Debugf("streams %+v", c.clientStreams)
+				c.clientStreams.Range(func(str, _ interface{}) bool {
+					if str.(quic.Stream).StreamID() >= f.StreamID {
+						str.(quic.Stream).SetDeadline(time.Now().Add(-42 * time.Second))
```

Comment:
Where do 42 seconds come from?

Reply:
Just a time in the past as was suggested. Is there a preferred way to do this?

Reply:
Actually I found a better way to handle this entirely